docs/proposals/hpa-custom-metrics.md

```diff
@@ -0,0 +1,162 @@
+to an entire namespace (e.g. queue length) and scale based on that.
+Additionally, until Heapster has the capability to track metrics from service
+objects, RCs, etc, namespace-level metrics can be used with push metrics to
+"fake" this type of metrics.  Metrics associated with services could be
+extracted from a load balancer or reverse proxy, for instance.
+
+Heapster will, most likely, support metrics associated with services,
+replicationcontrollers, etc, in the future, so it becomes adventageous to be
+able to refer to metrics associated with those objects as well.
+
+Proposed New Design
+-------------------
+
+```go
+type SourceType string
+var (
+    SourceTypeNamespace SourceType = "namespace"  // the current namespace
+    SourceTypePod SourceType = "pod"  // each pod that would be considered
+    SourceTypeController SourceType = "controller" // the target of the HPA
+)
+
+type MetricSource struct {
+    // Draw from the current namespace, pod, or controller
+    CurrentSource SourceType `json:"current,omitempty`
+
+    // Draw from an object in the current namespace (not currently supported
+    // by Heapster)
+    SourceRef *CrossVersionObjectReference `json:"object,omitempty"`
+}
+
+type MetricTarget struct {
+    // users could either specify
+    // `custom: my-metric` or `name: custom/my-metric`
+    // builtin metrics would be specified with `name: cpu/usage`
+    CustomName string `json:"custom,omitempty"`
+    BuiltinName string `json:"name,omitempty"`
+
+    TargetValue resource.Quantity `json:"targetValue"`
+
+    // Specifies the target object
+    // (missing is equivalent to {"current": "pod"})
+    SourceObject *MetricSource `json:"from,omitempty"`
+}
+
+type HorizontalPodAutoscalerSpec struct {
+    ScaleTargetRef CrossVersionObjectReference `json:"scaleTargetRef"`
+    MinReplicas *int32 `json:"minReplicas,omitempty"`
+    MaxReplicas int32 `json:"maxReplicas"`
+    Metrics []MetricTarget `json:"metrics"`
+}
+```
+
+This design allows us to refer to both custom metrics as well as built-in
+metrics, and allows referring to metrics from the current namespace, metrics
+from each pod individually (the default, and current setup), as well as metrics
+associated with other objects in the namespace.
+
+An HPA under this design would look like:
+
+```yaml
+apiVersion: autoscaling/v2
+kind: HorizontalPodAutoscaler
+metadata:
+    ...
+spec:
+    scaleTargetRef:
+        ...
+    maxReplicas: 10
+    metrics:
+    - name: memory/usage
+      targetValue: 256Mi
+    - custom: my-pod-metric
+      targetValue: 20
+    - custom: special-namespace-app-metric
+      from:
+        current: "namespace"
+    - custom: special-service-metric
+      from:
+        object:
+            kind: Service
+            apiVersion: v1
+            name: "my-service"
+```
+
+This would result in the following queries to Heapster (or a similar API):
+
+- `/api/v1/model/namespace/$NS/pod-list/$POD1,$POD2/metrics/memory/usage`
```

Comment:
yeah, that's how the Heapster model API currently works -- the HPA passes the list of pod names it wants metrics for, and gets back the results (the new metrics API doesn't work quite the same, but the decision was made not to support custom metrics in it for the time being).